### PR TITLE
add hal API for getting debug and profile deivce information

### PR DESCRIPTION
**Pull request by tianhaoz95**
This API will get the part of device information that debugger and profile need from `scan.h`.

I also plan to add it to xbutil dump so that GUI can make use of it when doing profiling or debugging.

The information will look like this in a host code:
```bash
2019-01-03 09:03:24,941 INFO [default] 	 Showing device[0] info:
2019-01-03 09:03:24,941 INFO [default] 		 Device Index: 0
2019-01-03 09:03:24,941 INFO [default] 		 User Instance: 129
2019-01-03 09:03:24,941 INFO [default] 		 NIFD Instance: 0
2019-01-03 09:03:24,941 INFO [default] 		 Management Instance: 1024
2019-01-03 09:03:24,941 INFO [default] 		 Device Name: /dev/dri/renderD129
2019-01-03 09:03:24,941 INFO [default] 		 Sysfs Name: /sys/bus/pci/devices/0000:04:00.1
2019-01-03 09:03:24,941 INFO [default] 		 NIFD Name: /dev/nifd0
```

### CODE REVIEW

**Comment by sonals** [comment]
ok to test

File: src/runtime_src/driver/include/xclperf.h [hunk 1]
```diff
@@ -413,4 +413,26 @@ typedef struct {
...
 
+#define DRIVER_NAME_ROOT "/dev"
+#define DEVICE_PREFIX "/dri/renderD"
+#define NIFD_PREFIX "/nifd"
+#define SYSFS_NAME_ROOT "/sys/bus/pci/devices/"
+
+/**
+ * \brief data structure for querying device info
+ * 
+ * device_type: 0 for sw_emu, 1 for hw_emu, 2 for xbb, 3 for aws
```
[the remark this answers is not in the document]
Comment (tianhaoz95) [reply src/runtime_src/driver/include/xclperf.h]: I changed it to an enum defined in the same header file.

```cpp
enum DeviceType {
  SW_EMU = 0,
  HW_EMU = 1,
  XBB = 2,
  AWS = 3
};

typedef struct {
  DeviceType device_type;
  unsigned int device_index;
  ...
}
```

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp [hunk 2]
```diff
@@ -961,6 +962,25 @@ namespace xocl {
...
+    uint16_t nifd_instance = 0;
+    std::string device_name = std::string(DRIVER_NAME_ROOT) + std::string(DEVICE_PREFIX) + std::to_string(user_instance);
+    std::string nifd_name = std::string(DRIVER_NAME_ROOT) + std::string(NIFD_PREFIX) + std::to_string(nifd_instance);
+    std::string sysfs_name = SYSFS_NAME_ROOT + dev->user->sysfs_name;
+    info.device_type = 2;
+    info.device_index = mBoardNumber;
+    info.user_instance = user_instance;
+    info.mgmt_instance = mgmt_instance;
+    info.nifd_instance = nifd_instance;
+    strcpy(info.device_name, device_name.c_str());
```
[the remark this answers is not in the document]
Comment (tianhaoz95) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: changed to:

```cpp
#define MAX_NAME_LEN 256

  char device_name[MAX_NAME_LEN];
  char sysfs_name[MAX_NAME_LEN];
  char nifd_name[MAX_NAME_LEN];
```

```cpp
strncpy(info.nifd_name, nifd_name.c_str(), MAX_NAME_LEN);
```

**Comment by tianhaoz95** [comment]
I just pushed the suggested changes. Can someone rebuild this when the server is ready?

**Comment by maxzhen** [comment]
ok to test

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: strncpy is actually more dangerous that strcpy.   If MAX_NAME_LEN is reached before the entire array src is copied, the dst array is *not* null-terminated.

Never use strncpy without checking if length was exceeded and if so manually null terminate the dst.   At least strcpy copies the entire null terminated str string while possibly exceed the boundary.

[the remark this answers is not in the document]
Comment (maxzhen) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: Use strlcpy(), then :-).

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp [hunk 3]
```diff
@@ -961,6 +962,25 @@ namespace xocl {
...
+    std::string nifd_name = std::string(DRIVER_NAME_ROOT) + std::string(NIFD_PREFIX) + std::to_string(nifd_instance);
+    std::string sysfs_name = SYSFS_NAME_ROOT + dev->user->sysfs_name;
+    info.device_type = DeviceType::XBB;
+    info.device_index = mBoardNumber;
+    info.user_instance = user_instance;
+    info.mgmt_instance = mgmt_instance;
+    info.nifd_instance = nifd_instance;
+    strncpy(info.device_name, device_name.c_str(), MAX_NAME_LEN);
+    strncpy(info.sysfs_name, sysfs_name.c_str(), MAX_NAME_LEN);
+    strncpy(info.nifd_name, nifd_name.c_str(), MAX_NAME_LEN);
```
Comment (stsoe) [inline src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: Please check the return value of strncpy and possibly manually null terminate.

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: strlcpy is unfortunately not standard.   Not sure if you can count on it being available.

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp (hunk 3 above)
Comment (maxzhen) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: or just blindly put a 0 at the end of the char array after calling strncpy().

**Comment by abalasa** [comment]
retest this please

File: src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp (hunk 3 above)
Comment (tianhaoz95) [reply src/runtime_src/driver/xclng/xrt/user_gem/perf.cpp]: I just pushed a change to forced all char array to have a null at the end:

```cpp
info.device_name[MAX_NAME_LEN-1] = '\0';
info.sysfs_name[MAX_NAME_LEN-1] = '\0';
info.nifd_name[MAX_NAME_LEN-1] = '\0';
```

**Comment by tianhaoz95** [comment]
just resolved a conflict introduced by enum clean up. Please build again

File: src/runtime_src/driver/include/xclperf.h [hunk 4]
```diff
@@ -413,4 +413,37 @@ typedef struct {
...
+ * nifd driver.
+ */
+typedef struct {
+  DeviceType device_type;
+  unsigned int device_index;
+  unsigned int mgmt_instance;
+  unsigned int user_instance;
+  unsigned int nifd_instance;
+  char device_name[MAX_NAME_LEN];
+  char sysfs_name[MAX_NAME_LEN];
```
Comment (maxzhen) [inline src/runtime_src/driver/include/xclperf.h]: Do you still need sysfs_name here?

Comment (tianhaoz95) [reply src/runtime_src/driver/include/xclperf.h]: I just pushed a change to remove the sysfs from this API @maxzhen

File: src/runtime_src/driver/include/xclhal2.h [hunk 5]
```diff
@@ -1176,6 +1176,33 @@ XCL_DRIVER_DLLESPEC uint32_t xclPerfMonGetTraceCount(xclDeviceHandle handle, enu
...
+ * Warning: These APIs are experimental and can be 
+ * changed or removed at any time. They should only 
+ * be used by debug and profile code.
+ * 
+ * @param handle the device handle
+ * @param info the xclDebugProfileDeviceInfo 
+ * structure that this API will fill in as 
+ * result
+ */
+XCL_DRIVER_DLLESPEC int xclGetDebugProfileDeviceInfo(xclDeviceHandle handle, xclDebugProfileDeviceInfo& info);
```
Comment (maxzhen) [inline src/runtime_src/driver/include/xclhal2.h]: Please do not use reference, this header file is also used by C code.

Comment (tianhaoz95-xilinx) [reply src/runtime_src/driver/include/xclhal2.h]: I just changed the reference to a pointer. It failed a check saying cannot fetch origin. I am asking Prema for some solution.

**Comment by tianhaoz95-xilinx** [comment]
The error message of this [build](http://xcoengvm229019:8080/job/XBB/job/XRT/job/XRT_PR_Builds/470/console) says:

```
ERROR: Error fetching remote repo 'origin'
Finished: FAILURE
```

It seems to be a infra failure. Can someone spin a rebuild to see if that's the case?

@gbuildx

**Comment by tianhaoz95** [comment]
This PR is now ready @maxzhen. The last build was an internet connection issue (could not clone).

**Pull request merged**